### ISSUE TITLE
Gradient exploded!

### ISSUE THREAD

**Issue by shivSD** [wizyoung/YOLOv3_TensorFlow#56]
I'm training with my own training set (there is a lot of variation in the data set), every time I run the training there is 'Gradient exploded!  Please train again and you may need to modify some parameters'. I did change the Learning rate(lowered) and also increased the warmup epoch 20-40 and still getting this issue. Is there any other parameter I need to change. here is my arg.py
I appreciate your help in advance.

## Training releated numbers
batch_size = 64
img_size = [416, 416]  # Images will be resized to `img_size` and fed to the network, size format: [width, height]
total_epoches = 500
train_evaluation_step = 500  # Evaluate on the training batch after some steps.
val_evaluation_epoch = 100  # Evaluate on the whole validation dataset after some steps. Set to None to evaluate every epoch.
save_epoch = 100  # Save the model after some epochs.
batch_norm_decay = 0.99  # decay in bn ops
weight_decay = 5e-4  # l2 weight decay
global_step = 0  # used when resuming training

## tf.data parameters
num_threads = 10  # Number of threads for image processing used in tf.data pipeline.
prefetech_buffer = 5  # Prefetech_buffer used in tf.data pipeline.

## Learning rate and optimizer
optimizer_name = 'adam'  # Chosen from [sgd, momentum, adam, rmsprop]
save_optimizer = True  # Whether to save the optimizer parameters into the checkpoint file.
learning_rate_init = 1e-6
lr_type = 'piecewise'  # Chosen from [fixed, exponential, cosine_decay, cosine_decay_restart, piecewise]
lr_decay_epoch = 5  # Epochs after which learning rate decays. Int or float. Used when chosen `exponential` and `cosine_decay_restart` lr_type.
lr_decay_factor = 0.96  # The learning rate decay factor. Used when chosen `exponential` lr_type.
lr_lower_bound = 1e-6  # The minimum learning rate.
# piecewise params
pw_boundaries = [60, 80]  # epoch based boundaries
pw_values = [learning_rate_init, 3e-5, 1e-4]

## Load and finetune
restore_part = ['yolov3/darknet53_body']
update_part = ['yolov3/yolov3_head']

## other training strategies
multi_scale_train = False  # Whether to apply multi-scale training strategy. Image size varies from [320, 320] to [640, 640] by default.
use_label_smooth = False # Whether to use class label smoothing strategy.
use_focal_loss = False  # Whether to apply focal loss on the conf loss.
use_mix_up = False  # Whether to use mix up data augmentation strategy. 
use_warm_up = True  # whether to use warm up strategy to prevent from gradient exploding.
warm_up_epoch = 30  # Warm up training epoches. Set to a larger value if gradient explodes.

**Comment by wizyoung** [comment]
That's weird. Maybe the anchors are not appropriate?

**Comment by fdy200533** [comment]
The same problem，loss nan happens too frequently

**Comment by henbucuoshanghai** [comment]
the same to you      exploded  mang times???   why
?   there are many parameters?  change them all???

**Comment by shivSD** [comment]
@wizyoung, i tried coco anchor and anchor generated from training set. Both are having the same issue.

**Comment by henbucuoshanghai** [comment]
has any one sloved it ????????????

**Comment by olcaybuyan** [comment]
Gradient clipping seems to solve it. See this post on how to apply:
https://stackoverflow.com/questions/36498127/how-to-apply-gradient-clipping-in-tensorflow

**Comment by wizyoung** [comment]
(1) try the two-stage training strategy in the readme file.
(2) If you want to finetune the whole model, try using momentum optimizer instead of the default adam

**Comment by wizyoung** [comment]
By the way, the loss function is updated (the previous version is buggy)

**Comment by henbucuoshanghai** [comment]
powerful




------------------ Original message ------------------
From: "Wizyoung"; 
Sendtime: Tuesday, Jun 25, 2019 1:01 AM
To: "wizyoung/YOLOv3_TensorFlow"; 
Cc: "很不错学生"<1119006913@qq.com>; "Comment"; 
Subject: Re: [wizyoung/YOLOv3_TensorFlow] Gradient exploded! (#56)




By the way, the loss function is updated (the previous version is buggy)
